### PR TITLE
Simplify GetNodeExecutionStatus

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

# TL;DR
For ev 0, parentnodeid is not important except for branch nodes. In case of dynamic workflow caching this was not set correctly. This PR changes the behavior and also streamlines the code block

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
For event version-0 the dynamic array task did not generate the right node ids and relationship. This change fixes it

## Tracking Issue

fixes https://github.com/flyteorg/flyte/issues/1236

### CODE REVIEW

File: pkg/apis/flyteworkflow/v1alpha1/node_status.go [hunk 1]
```diff
@@ -517,62 +518,53 @@ func (in NodeStatus) GetTaskNodeStatus() ExecutableTaskNodeStatus {
...
-			if err != nil {
-				logger.Errorf(ctx, "Failed to construct output dir for node [%v]", id)
-				return n
-			}
+		n.SetDataDir(dataDir)
+	}
 
-			n.SetOutputDir(outputDir)
+	if len(n.GetOutputDir()) == 0 {
+		outputDir, err := in.DataReferenceConstructor.ConstructReference(ctx, n.GetDataDir(), strconv.FormatUint(uint64(n.Attempts), 10))
```
Comment (katrogan) [inline pkg/apis/flyteworkflow/v1alpha1/node_status.go:533]: why is this using n.GetDataDir to set the outputDir and above in line 524 the OutputDir to set the dataDir. This seems backwards

File: pkg/controller/nodes/dynamic/dynamic_workflow.go [hunk 2]
```diff
@@ -171,27 +201,53 @@ func (d dynamicNodeTaskNodeHandler) buildContextualDynamicWorkflow(ctx context.C
...
+		return workflowContext, err
+	}
+
+	if err := f.Cache(ctx, dynamicWf, closure); err != nil {
+		logger.Errorf(ctx, "Failed to cache Dynamic workflow [%s]", err.Error())
+	}
+
+	// The current node would end up becoming the parent for the dynamic task nodes.
+	// This is done to track the lineage. For level zero, the CreateParentInfo will return nil
+	newParentInfo, err := node_common.CreateParentInfo(nCtx.ExecutionContext().GetParentInfo(), nCtx.NodeID(), nCtx.CurrentAttempt())
```
Comment (katrogan) [inline pkg/controller/nodes/dynamic/dynamic_workflow.go:215]: do we not need to create the new parent info in the cache read case too?

File: pkg/controller/nodes/dynamic/dynamic_workflow.go [hunk 3]
```diff
@@ -171,27 +201,53 @@ func (d dynamicNodeTaskNodeHandler) buildContextualDynamicWorkflow(ctx context.C
...
+	closure, dynamicWf, workflowContext, err := d.buildDynamicWorkflow(ctx, nCtx, djSpec, dynamicNodeStatus)
+	if err != nil {
+		return workflowContext, err
+	}
+
+	if err := f.Cache(ctx, dynamicWf, closure); err != nil {
+		logger.Errorf(ctx, "Failed to cache Dynamic workflow [%s]", err.Error())
+	}
+
+	// The current node would end up becoming the parent for the dynamic task nodes.
```
Comment (katrogan) [inline pkg/controller/nodes/dynamic/dynamic_workflow.go:213]: do you need to call setEphemeralNodeExecutionStatusAttributes here too?

File: pkg/apis/flyteworkflow/v1alpha1/node_status.go (hunk 1 above)
Comment (kumare3) [reply pkg/apis/flyteworkflow/v1alpha1/node_status.go:533]: since the datadir is already set, output dir is just a derivative from that

File: pkg/controller/nodes/dynamic/dynamic_workflow.go (hunk 3 above)
Comment (kumare3) [reply pkg/controller/nodes/dynamic/dynamic_workflow.go:213]: in the case of the cache, sadly, GetNodeExecutionStatus has a side-effect that sets those attributes

**Review by kumare3 — APPROVED** [review]
